### PR TITLE
[Azure Search] Add initial search rankings based on download counts

### PR DESCRIPTION
**Pull request by shishirx34**
I want this to be in so that a scoring profile is applied for any index built, the index can then be manipulated from REST API for rankings.

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/IndexBuilder.cs [hunk 1]
```diff
@@ -103,9 +104,9 @@ private Index InitializeHijackIndex()
                 _options.Value.HijackIndexName);
         }
 
-        private Index InitializeIndex<TDocument>(string name)
+        private Index InitializeIndex<TDocument>(string name, bool addScoringProfile = false)
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/IndexBuilder.cs]: >addScoringProfile [](start = 67, length = 17)

Let's be explicit for `InitializeHijackIndex`, no need for a default value here

File: src/NuGet.Services.AzureSearch/Analysis/DescriptionCustomAnalyzer.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@ public static class DescriptionAnalyzer
                 IdentifierCustomTokenFilter.Name,
                 TokenFilterName.Stopwords,
                 TokenFilterName.Lowercase,
+                TokenFilterName.Truncate
             });
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Analysis/DescriptionCustomAnalyzer.cs]: Could you add a TODO with the bug you opened: [https://github.com/NuGet/NuGetGallery/issues/7045](https://github.com/NuGet/NuGetGallery/issues/7045)

This is perfectly fine to unblock yourself, but it's not the final solution.

Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/Analysis/DescriptionCustomAnalyzer.cs]: Same with `PackageIdCustomAnalyzer`

---
In reply to: [276023764](https://github.com/NuGet/NuGet.Services.Metadata/pull/506#discussion_r276023764) [](ancestors = 276023764)

File: src/NuGet.Services.AzureSearch/Analysis/IdentifierCustomTokenFilter.cs [hunk 3]
```diff
@@ -16,6 +16,7 @@ public static class IdentifierCustomTokenFilter
 
         public static WordDelimiterTokenFilter Instance = new WordDelimiterTokenFilter(
             Name,
+            catenateAll: true,
             splitOnCaseChange: true,
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Analysis/IdentifierCustomTokenFilter.cs]: Please update this class's comments and the functional tests to reflect this change.

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs [hunk 4]
```diff
@@ -19,6 +19,7 @@ public static class SearchDocument
         [SerializePropertyNamesAsCamelCase]
         public class Full : AddFirst
         {
+            [IsFilterable]
             public long? TotalDownloadCount { get; set; }
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Models/SearchDocument.cs]: What's this for?

File: src/NuGet.Services.AzureSearch/ScoringProfiles/DownloadCountBoosterProfile.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+using NuGet.Services.AzureSearch.SearchService;
+
+namespace NuGet.Services.AzureSearch.ScoringProfiles
+{
+    public class DownloadCountBoosterProfile
+    {
+        public const string Name = "download_count_scoring_profile";
+
+        public static readonly ScoringProfile Instance = new ScoringProfile(
+                Name,
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/ScoringProfiles/DownloadCountBoosterProfile.cs]: >         [](start = 8, length = 8)

You have one tab too many on lines 15-32 :)

File: src/NuGet.Services.AzureSearch/Analysis/IdentifierCustomTokenFilter.cs (hunk 3 above)
Comment (shishirx34) [reply src/NuGet.Services.AzureSearch/Analysis/IdentifierCustomTokenFilter.cs]: I didn't mean to add it just yet. removing it, will thoroughly test and add appropriate testing for this case.

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs (hunk 4 above)
Comment (shishirx34) [reply src/NuGet.Services.AzureSearch/Models/SearchDocument.cs]: the scoring profile can only be added to a field that is "filterable"

File: src/NuGet.Services.AzureSearch/Analysis/DescriptionCustomAnalyzer.cs (hunk 2 above)
Comment (shishirx34) [reply src/NuGet.Services.AzureSearch/Analysis/DescriptionCustomAnalyzer.cs]: I think the spacing split for tokenization is only applicable to the description analyzer, not to the package id analyzer.

**Pull request merged**